app/api/ada/hardware-wallet/connectTrezorAdaWallet.js

```diff
@@ -0,0 +1,232 @@
+// @flow
+
+// Handles Connecting Trezor with Yoroi that follow the v2 addressing scheme (bip44)
+
+import _ from 'lodash';
+import { Wallet } from 'rust-cardano-crypto';
+import { getResultOrFail } from '../lib/cardanoCrypto/cryptoUtils';
+import {
+  saveAsAdaAddresses,
+  newAdaAddress
+} from '../adaAddress';
+import {
+  checkAddressesInUse,
+} from '../lib/yoroi-backend-api';
+import type {
+  AdaWallet,
+  AdaHardwareWalletParams
+} from '../adaTypes';
+import {
+  Logger,
+  stringifyError
+} from '../../../utils/logging';
+import { saveCryptoAccount, saveAdaWallet } from '../adaLocalStorage';
+import { createAdaHardwareWallet } from '../adaWallet';
+import { createHardwareWalletAccount } from '../adaAccount';
+import type { ConfigType } from '../../../../config/config-types';
+
+type AddressInfo = { address: string, isUsed: boolean, index: number };
+
+declare var CONFIG: ConfigType;
+const addressScanSize = CONFIG.app.addressScanSize;
+const addressRequestSize = CONFIG.app.addressRequestSize;
+
+/** Creates a new/empty ada wallet
+ * Fetches all related addresses
+ * Caches the fetched address results locally
+ * @returns a new AdaWallet
+ */
+export async function connectTrezorAdaWallet({
+  walletInitData
+}: AdaHardwareWalletParams): Promise<AdaWallet> {
+  try {
+    Logger.debug('connectTrezorAdaWallet::connectTrezorAdaWallet called');
+
+    // create ada wallet object for hardware wallet
+    const [adaWallet] = createAdaHardwareWallet({ walletInitData });
+    // create crypto account object for hardware wallet
+    // eslint-disable-next-line max-len
+    const cryptoAccount = createHardwareWalletAccount(walletInitData.cwHardwareInfo.publicMasterKey);
+
+    // fetch all addresses
+    // Note: addresses may be skipped if generated by a wallet that doesn't follow bip44
+
+    // fetch all External addresses
+    const externalAddressesToSave =
+      await _discoverAllAddressesFrom(cryptoAccount, 'External', -1, addressScanSize, addressRequestSize);
+
+    // fetch all Internal addresses
+    const internalAddressesToSave =
+      await _discoverAllAddressesFrom(cryptoAccount, 'Internal', -1, addressScanSize, addressRequestSize);
+
+    // store wallet related addresses to lovefieldDatabase
+    if (externalAddressesToSave.length !== 0 || internalAddressesToSave.length !== 0) {
+      await Promise.all([
+        saveAsAdaAddresses(cryptoAccount, externalAddressesToSave, 'External'),
+        saveAsAdaAddresses(cryptoAccount, internalAddressesToSave, 'Internal')
+      ]);
+    } else {
+      // no related addresses found, give it a new address and save it to lovefieldDatabase
+      await newAdaAddress(cryptoAccount, [], 'External');
+    }
+
+    // save crypto account to the local storage
+    saveCryptoAccount(cryptoAccount);
+
+    // save ada wallet to the local storage
+    saveAdaWallet(adaWallet);
+
+    // It's a success, we are done
```

Comment:
this should be on the top

Reply:
like this is what this function does

Reply:
https://github.com/Emurgo/yoroi-frontend/commit/cc95c0f729587dbfaca9b3c7fbdd8f11b145cb5d